amr-wind/equation_systems/icns/source_terms/DragForcing.cpp

```diff
@@ -0,0 +1,179 @@
+    const unsigned vsize = m_device_vel_ht.size();
+    const auto& dt = m_time.delta_t();
+    bool is_laminar = m_is_laminar;
+    const amrex::Real scale_factor = (dx[2] < 1.0) ? 1.0 : 1.0 / dx[2];
+    const amrex::Real Cd =
+        (is_laminar && dx[2] < 1) ? dragCoefficient : dragCoefficient / dx[2];
+    amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+        const amrex::Real x = prob_lo[0] + (i + 0.5) * dx[0];
+        const amrex::Real y = prob_lo[1] + (j + 0.5) * dx[1];
+        const amrex::Real z = prob_lo[2] + (k + 0.5) * dx[2];
+        amrex::Real xstart_damping = 0;
+        amrex::Real ystart_damping = 0;
+        amrex::Real xend_damping = 0;
+        amrex::Real yend_damping = 0;
+        amrex::Real xi_end = (x - start_east) / (prob_hi[0] - start_east);
+        amrex::Real xi_start = (start_west - x) / (start_west - prob_lo[0]);
+        xi_start = sponge_west * std::max(xi_start, 0.0);
+        xi_end = sponge_east * std::max(xi_end, 0.0);
+        xstart_damping = sponge_west * sponge_strength * xi_start * xi_start;
+        xend_damping = sponge_east * sponge_strength * xi_end * xi_end;
+        amrex::Real yi_end = (y - start_north) / (prob_hi[1] - start_north);
+        amrex::Real yi_start = (start_south - y) / (start_south - prob_lo[1]);
+        yi_start = sponge_south * std::max(yi_start, 0.0);
+        yi_end = sponge_north * std::max(yi_end, 0.0);
+        ystart_damping = sponge_strength * yi_start * yi_start;
+        yend_damping = sponge_strength * yi_end * yi_end;
+        amrex::Real spongeVelX = 0.0;
+        amrex::Real spongeVelY = 0.0;
+        amrex::Real spongeVelZ = 0.0;
+        amrex::Real residual = 1000;
+        amrex::Real height_error = 0.0;
+        for (unsigned ii = 0; ii < vsize; ++ii) {
+            height_error = std::abs(z - device_vel_ht[ii]);
+            if (height_error < residual) {
+                residual = height_error;
+                const unsigned ix = 3 * ii;
+                const unsigned iy = 3 * ii + 1;
+                const unsigned iz = 3 * ii + 2;
+                spongeVelX = device_vel_vals[ix];
+                spongeVelY = device_vel_vals[iy];
+                spongeVelZ = device_vel_vals[iz];
+            }
+        }
+        amrex::Real Dxz = 0.0;
+        amrex::Real Dyz = 0.0;
+        amrex::Real bcforcing_x = 0;
+        amrex::Real bcforcing_y = 0;
+        const amrex::Real ux1 = vel(i, j, k, 0);
+        const amrex::Real uy1 = vel(i, j, k, 1);
+        const amrex::Real uz1 = vel(i, j, k, 2);
+        const amrex::Real m = std::sqrt(ux1 * ux1 + uy1 * uy1 + uz1 * uz1);
+        if (drag(i, j, k) == 1 && (!is_laminar)) {
+            const amrex::Real ux2 = vel(i, j, k + 1, 0);
+            const amrex::Real uy2 = vel(i, j, k + 1, 1);
+            const amrex::Real m2 = std::sqrt(ux2 * ux2 + uy2 * uy2);
+            const amrex::Real kappa = 0.41;
+            const amrex::Real z0 = std::max(terrainz0(i, j, k), 1e-4);
```

Comment:
does the minimum value of z0 have some significance? Seems like it should be specified in a more obvious place or should be an input argument

Reply:
It is possible to have a -999 in the input file for water bodies sometimes. I can change it to a minimum roughness length variable.